go/store/datas/pull/pull_chunk_tracker.go

```diff
@@ -0,0 +1,261 @@
+		reqCh:       make(chan *trackerGetAbsentReq),
+	}
+	ret.seen.InsertAll(initial)
+	ret.wg.Add(1)
+	go func() {
+		defer ret.wg.Done()
+		ret.thread(initial)
+	}()
+	return ret
+}
+
+func (t *PullChunkTracker) Seen(h hash.Hash) {
+	if !t.seen.Has(h) {
+		t.seen.Insert(h)
+		t.addUnchecked(h)
+	}
+}
+
+func (t *PullChunkTracker) Close() {
+	close(t.uncheckedCh)
+	t.wg.Wait()
+}
+
+func (t *PullChunkTracker) addUnchecked(h hash.Hash) {
+	select {
+	case t.uncheckedCh <- h:
+	case <-t.ctx.Done():
+	}
+}
+
+func (t *PullChunkTracker) GetChunksToFetch() (hash.HashSet, bool, error) {
+	var req trackerGetAbsentReq
+	req.ready = make(chan struct{})
+
+	select {
+	case t.reqCh <- &req:
+	case <-t.ctx.Done():
+		return nil, false, t.ctx.Err()
+	}
+
+	select {
+	case <-req.ready:
+	case <-t.ctx.Done():
+		return nil, false, t.ctx.Err()
+	}
+
+	return req.hs, req.ok, req.err
+}
+
+func (t *PullChunkTracker) thread(initial hash.HashSet) {
+	doneCh := make(chan struct{})
+	hasManyReqCh := make(chan trackerHasManyReq)
+	hasManyRespCh := make(chan trackerHasManyResp)
+
+	var wg sync.WaitGroup
+	wg.Add(t.cfg.HasManyThreadCount)
+
+	for i := 0; i < t.cfg.HasManyThreadCount; i++ {
+		go func() {
+			defer wg.Done()
+			hasManyThread(t.ctx, t.cfg.Haser, hasManyReqCh, hasManyRespCh, doneCh)
+		}()
+	}
+
+	defer func() {
+		close(doneCh)
+		wg.Wait()
+	}()
+
+	unchecked := make([]hash.HashSet, 0)
+	absent := make([]hash.HashSet, 0)
+
+	var err error
+	outstanding := 0
+
+	if len(initial) > 0 {
+		unchecked = append(unchecked, initial)
+		outstanding += 1
+	}
+
+	for {
+		var thisReqCh = t.reqCh
+		if outstanding != 0 && len(absent) == 0 {
+			// If we are waiting for a HasMany response and we don't currently have any
+			// absent addresses to return, block any absent requests.
+			thisReqCh = nil
+		}
+
+		var thisHasManyReqCh chan trackerHasManyReq
+		var hasManyReq trackerHasManyReq
+		if len(unchecked) > 0 {
+			hasManyReq.hs = unchecked[0]
+			thisHasManyReqCh = hasManyReqCh
+		}
+
+		select {
+		case h, ok := <-t.uncheckedCh:
+			if !ok {
+				return
+			}
+			if len(unchecked) == 0 || len(unchecked[len(unchecked)-1]) >= t.cfg.BatchSize {
+				outstanding += 1
+				unchecked = append(unchecked, make(hash.HashSet))
+			}
+			unchecked[len(unchecked)-1].Insert(h)
+		case resp := <-hasManyRespCh:
+			outstanding -= 1
+			if resp.err != nil {
+				err = errors.Join(err, resp.err)
+			} else if len(resp.hs) > 0 {
+				absent = append(absent, resp.hs)
+			}
+		case thisHasManyReqCh <- hasManyReq:
+			copy(unchecked[:], unchecked[1:])
+			if len(unchecked) > 1 {
+				unchecked[len(unchecked)-1] = nil
+			}
+			unchecked = unchecked[:len(unchecked)-1]
+		case req := <-thisReqCh:
+			if err != nil {
+				req.err = err
+				close(req.ready)
+				err = nil
+			} else if len(absent) == 0 {
+				req.ok = false
+				close(req.ready)
+			} else {
+				req.ok = true
+				req.hs = absent[0]
+				var i int
+				for i = 1; i < len(absent); i++ {
```

Comment:
A little confused here. If the next absent batch doesn't go beyond total batch size, we add the whole thing, otherwise break out of the loop. I'm not sure what the sizes of the batches are, but I'm assuming the the absent batch sizes are considerable smaller. Is it worth it to add a subset of absent[i] so that we get a full batch?

Reply:
Rounded out the last batch.